### PR TITLE
sleep: add SleepMemMode= setting for configuring /sys/power/mem_sleep

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

File: src/shared/sleep-config.h [hunk 1]
```diff
@@ -39,6 +41,13 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(SleepConfig*, sleep_config_free);
 
 int parse_sleep_config(SleepConfig **sleep_config);
 
+static inline bool SLEEP_NEEDS_MEM_SLEEP(const SleepConfig *sc, SleepOperation operation) {
+        assert(sc);
+        assert(operation >= 0 && operation < _SLEEP_OPERATION_CONFIG_MAX);
+        return strv_contains(sc->states[operation], "mem") ||
+               strv_contains(sc->modes[operation], "suspend");
```
Comment (yuwata) [inline src/shared/sleep-config.h:53]: "hybrid-sleep" ? Or, documentation is wrong?

**Review by yuwata — APPROVED** [review]
LGTM. Except for the below.

File: src/shared/sleep-config.h (hunk 1 above)
Comment (YHNdnzj) [reply src/shared/sleep-config.h:53]: "suspend" in /sys/power/disk means to put the system into suspension after writing hibernation image, i.e. hybrid-sleep.

Comment (yuwata) [reply src/shared/sleep-config.h:53]: OK. Could you add a short comment about that in the code?

Comment (yuwata) [reply src/shared/sleep-config.h:53]: Ah, that's not `sc->states`, but `sc->modes`.
Still adding comment is welcome.

File: src/shared/sleep-config.c [hunk 2]
```diff
@@ -140,6 +142,8 @@ int parse_sleep_config(SleepConfig **ret) {
                 { "Sleep", "HybridSleepState",          config_parse_warn_compat, DISABLED_LEGACY, NULL                         },
                 { "Sleep", "HybridSleepMode",           config_parse_warn_compat, DISABLED_LEGACY, NULL                         },
 
+                { "Sleep", "SleepMemMode",              config_parse_sleep_mode,  0,               &sc->mem_modes               },
```
Comment (poettering) [inline src/shared/sleep-config.c:145]: this is cryptic enough, let's not needlessly abbreviate here.

MemorySleepMode=?

File: man/systemd-sleep.conf.xml [hunk 3]
```diff
@@ -195,6 +195,24 @@
...
 
+      <varlistentry>
+        <term><varname>SleepMemMode=</varname></term>
+
+        <listitem><para>The string to be written to <filename>/sys/power/mem_sleep</filename>
+        when <option>SuspendState=mem</option> or <command>hybrid-sleep</command> is used.
+        More than one value can be specified by separating multiple values with whitespace. They will be
+        tried in turn, until one is written without error. If none of the writes succeed, the operation will
+        be aborted. Defaults to empty, i.e. the kernel default or kernel command line option
+        <varname>mem_sleep_default=</varname> is respected.</para>
```
Comment (poettering) [inline man/systemd-sleep.conf.xml:206]: this stuff is all so cryptic. we should probably start listing common values here at least,

**Review by poettering — COMMENTED** [review]
some post-merge comments

File: src/shared/sleep-config.c (hunk 2 above)
Comment (YHNdnzj) [reply src/shared/sleep-config.c:145]: -> #32125

File: man/systemd-sleep.conf.xml (hunk 3 above)
Comment (YHNdnzj) [reply man/systemd-sleep.conf.xml:206]: The kernel documentation has all possible values? And this among all sleep options varies most depending on hardware...

**Pull request merged**